import osimport pandas as pdimport gzip#Set working directorty to CHD cancer rootchdcan_path = "/Users/jon/Documents/data/chdcan/chd_cancer"os.chdir(chdcan_path)#File paths relative to CHD cancer rootLG_panel = "cangenLG.txt"cancer_genes = "raw_data/genes_list/cosmic/Cosmic_Gene_Census_all_18072023.tsv"cancer_cur = "processed_data/genes_list/cosmic/py_curated.tsv"chd_genes = "raw_data/genes_list/chd/Gene_list_CHD_reclassification_20201111_FINAL.xlsx"chd_cur = "processed_data/genes_list/chd/py_curated.tsv"chd_vars = "chd_ukbb.sample_qced.variant_qced.maf1.vgs.15032023.ht.tsv.bgz"overlap_out = "processed_data/genes_list/overlap.tsv"filter_out = "processed_data/variants_table/cosmic/py_rare_variants_chdukbb.tsv"#Parsing gene list (merging cancer and CHD scripts from R)def parse_gene_list(data_in, data_out, mode):    if mode == "chd":        chd_genes = pd.read_excel(data_in, sheet_name=0, header=0)        chd_genes = chd_genes.rename(columns={            'genes CHD': 'gene', 'new ranking JB': 'tier_chd'})        chd_genes = chd_genes[['gene', 'tier_chd']]        chd_genes['tier_chd'] = chd_genes['tier_chd'].str.extract(r'CHD_(\d+)').fillna(-1).astype(int)        chd_genes.to_csv(data_out, sep='\t', index=False)    if mode == "cancer":        cancer_genes = pd.read_csv(data_in, sep='\t', header=0)        cancer_genes = cancer_genes[['Gene Symbol', 'Tier', 'Germline']].rename(columns={        'Gene Symbol': 'gene',        'Tier': 'tier_cancer',        'Germline': 'is_germline'    })        cancer_genes['is_germline'] = cancer_genes['is_germline'] == 'yes'        cancer_genes.to_csv(data_out, sep='\t', index=False)#Overlapping gene lists by merging on the gene columndef overlap_genes(a_in, b_in, data_out):    genes_a = pd.read_csv(a_in, sep='\t', header=0)    genes_b = pd.read_csv(b_in, sep='\t', header=0)    df_merged = pd.merge(genes_a, genes_b, on='gene', how='inner')    df_merged.to_csv(data_out, sep='\t', index=False)def filt_vars(var_data, gene_data, data_out, af=0.01):    if var_data.endswith("bgz"):        with gzip.open(var_data, 'rt') as f:            variants = pd.read_csv(f, sep='\t', header=0, dtype={24: str})    else:        variants = pd.read_csv(var_data, sep='\t', header=0, dtype={24: str})    genes = pd.read_csv(gene_data, sep='\t', header=0)['gene']    df_rare = variants[    ((variants['gnomad_genomes_af'] < af) |      (variants['gnomad_genomes_af'].isna())) &    ((variants['gnomAD_AF'] < af) |      (variants['gnomAD_AF'].isna())) &    (variants['SYMBOL'].isin(genes))]    df_rare.to_csv(data_out, sep='\t', index=False)#Runs/testing blockparse_gene_list(cancer_genes, cancer_cur, "cancer")parse_gene_list(chd_genes, chd_cur, "chd")overlap_genes(chd_cur, cancer_cur, overlap_out)filt_vars(chd_vars, overlap_out, filter_out)